staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -0,0 +1,375 @@
+}
+
+// ValidatingAdmissionPolicySpec is the specification of the desired behavior of the AdmissionPolicy.
+type ValidatingAdmissionPolicySpec struct {
+	//ParamSource specifies the kind of resources used to parameterize this policy which has to be a CRD.
+	// +optional
+	ParamSource ParamSource `json:"paramSource,omitempty" protobuf:"bytes,1,rep,name=paramSource"`
+
+	// MatchConstraints specifies what resources this policy is designed to validate.
+	// The AdmissionPolicy cares about a validation if it matches _any_ Constriant.
+	// However, in order to prevent clusters from being put into a unustable state that cannot be recoverd from via the API
+	// alidatingAdmissionPolicy cannot match ValidatingAdmissionPolicy/PolicyBinding/param resources.
+	// ValidatingWebhookConfiguration cannot match MutatingWebhookConfiguration or ValidatingAdmissionPolicy/PolicyBinding/param resources.
+	// +optional
+	MatchConstraints MatchResources `json:"matchConstraints,omitempty" protobuf:"bytes,2,rep,name=matchConstraints"`
+
+	// Validations contain CEL expressions which is used to apply the validation.
+	// Required.
+	Validations []Validation `json:"validations,omitempty" protobuf:"bytes,3,rep,name=validations"`
+
+	// FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+	// allowed values are Ignore or Fail. Defaults to Fail.
+	// +optional
+	FailurePolicy *FailurePolicyType `json:"failurePolicy,omitempty" protobuf:"bytes,4,opt,name=failurePolicy,casttype=FailurePolicyType"`
+}
+
+// ParamSource is a tuple of Group Kind and Version.
+type ParamSource struct {
+	// APIGroup is the API group the resources belong to.
+	// Required.
+	APIGroup string `json:"apiGroup,omitempty" protobuf:"bytes,1,rep,name=apiGroup"`
+
+	// APIVersion is the API version the resources belong to.
+	// Required.
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,2,rep,name=apiVersion"`
+
+	// APIKind is the API kind the resources belong to.
+	// Required.
+	APIKind string `json:"apiKind,omitempty" protobuf:"bytes,3,rep,name=apiKind"`
+}
+
+// Validation specified the CEL expression which is used to apply the validation.
+type Validation struct {
+	// The name of the validation rule.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the AdmissionReview type, organized into CEL variables as well as some other useful variables:
+	//
+	//'object'
+	//'oldObject'
+	//'review'
+	//'requestResource' (GVR)
+	//'resource' (GVR)
+	//'name'
+	//'namespace'
+	//'operation'
+	//'userInfo'
+	//'dryRun'
+	//'options'
+	//'config' - configuration data of the policy configuration being validated
+	// The `object` variable in the expression is bound to the resource this policy is designed to validate.
```

Comment:
Briefly explain each of these root variables? I expect that AdmissionReview has some documentation that can be used as a starting point.